File: metric.m
% delete(gcp)
% matlabpool 4
% % delete(gcp('nocreate'))
% myPool = parpool(2);
%% calculate the metrics of models 
% currentFolder = pwd;
metricsFolder = 'code4metric'
addpath(genpath(metricsFolder))

% folder_list={'snapshot-train_kldloss_iter_850000', 'snapshot-train_kldloss_withouteuc_iter_150000'}
folder_list={'snapshot-train_nss-kldloss_withouteuc_iter_150000', 'snapshot-train_nssloss_iter_550000', 'snapshot-train_nssloss_withouteuc_iter_100000'}

dens_dir = '/data/sunnycia/SaliencyDataset/Image/SALICON/DATA/train_val/val2014/density';
fixa_dir = '/data/sunnycia/SaliencyDataset/Image/SALICON/DATA/train_val/val2014/fixation';

for i=1:length(folder_list)
    folder = folder_list(i)

    sal_dir = char(fullfile('/data/sunnycia/SaliencyDataset/Image/SALICON/DATA/train_val/val2014/saliency/', folder))
    save_base_dir = '/data/sunnycia/saliency_on_videoset/Train/metric';

    s = dir(fullfile(sal_dir, '*.jpg'));
    d = dir(fullfile(dens_dir, '*.jpg'));
    f = dir(fullfile(fixa_dir, '*.mat'));

    saliencymap_path_list = {s.name};
    densitymap_path_list = {d.name};
    fixationmap_path_list = {f.name};
    saliencymap_path_list = sort(saliencymap_path_list);
    densitymap_path_list = sort(densitymap_path_list);
    fixationmap_path_list = sort(fixationmap_path_list);
    % disp(saliencymap_path_list);

    LengthFiles = length(saliencymap_path_list);
    saliency_score_CC = zeros(1,LengthFiles);
    saliency_score_SIM = zeros(1,LengthFiles);
    saliency_score_JUD = zeros(1,LengthFiles);
    saliency_score_BOR = zeros(1,LengthFiles);
    saliency_score_SAUC = zeros(1,LengthFiles);
    saliency_score_EMD = zeros(1,LengthFiles);
    saliency_score_KL = zeros(1,LengthFiles);
    saliency_score_NSS = zeros(1,LengthFiles);

    for j = 1 : LengthFiles
        sal_map_path = char(saliencymap_path_list(j))
        dens_map_path = char(densitymap_path_list(j))
        fix_map_path = char(fixationmap_path_list(j))
        [pathstr,sname,ext] = fileparts(sal_map_path);
        [pathstr,dname,ext] = fileparts(dens_map_path);
        [pathstr,fname,ext] = fileparts(fix_map_path);

        % sname, dname, fname;
        assert( strcmp(sname, dname)==1 && strcmp(dname, fname)==1)

        smap_path = fullfile(sal_dir, sal_map_path);
        density_path = fullfile(dens_dir, dens_map_path);
        fixation_path = fullfile(fixa_dir, fix_map_path);



        image_saliency = imread(smap_path);
        image_density = imread(density_path);
        load(fixation_path);
        image_fixation = fixation;
        

        %% CC %%
        %tic
        saliency_score_CC(j) = CC(image_saliency, image_density);
        %toc
        %% SIM %% 
        %tic
        saliency_score_SIM(j) = similarity(image_saliency, image_density);
        %toc
        
        %% AUCJUDD %%
        %tic
        saliency_score_JUD(j) = AUC_Judd(image_saliency, image_fixation, 0);
        %toc
        
        %% AUCBorji %%
        %tic
        saliency_score_BOR(j) = AUC_Borji(image_saliency, image_fixation);
        %toc
        
        %% SAUC %%
        %tic
        % saliency_score_SAUC(j) = AUC_shuffled(image_saliency, image_fixation, other_map);
        %toc
        
        %% EMD %%
        %tic
        % saliency_score_EMD(j) = EMD(image_saliency, image_density);
        % saliency_score_EMD(j) = emd_hat_mex(image_saliency, image_fixation, image_fixation);
        %toc
        %% KL %%
        %tic
        saliency_score_KL(j) = KLdiv(image_saliency, image_density);
        %toc
        
        %% NSS %%
        %tic
        saliency_score_NSS(j)=NSS(image_saliency, image_density);
        %toc
        % fprintf('Done for %s\n', smap_path);
    end
    saliency_score=[saliency_score_CC;saliency_score_SIM;
                    saliency_score_JUD;saliency_score_BOR;
                    saliency_score_SAUC;saliency_score_EMD;
                    saliency_score_KL;saliency_score_NSS;];

    % [pathstr,modelname,ext] = fileparts(sal_dir);
    save_name = strcat(folder, '.mat');
    save_path = fullfile(save_base_dir, save_name);
    save(char(save_path), 'saliency_score');
    fprintf('%s saved\n',save_path);

end

File: metric/metric.py
from utils.pymetric.metrics import *
import os
import glob
import scipy.io as scio
import cv2
from utils.common import *
import argparse
# from utils. import Colored
# metric_list = ['CC'color_print, 'SIM', 'AUC_JUD', 'AUC_BOR', 'SAUC', 'EMD', 'KLD', 'NSS']
# mask_code = [0, 0, 0, 0, ]
def get_arguments():
    parser = argparse.ArgumentParser()
    parser.add_argument('--debug', type=bool, default=False)
    parser.add_argument('--dsname', type=str, required=True)
    parser.add_argument('--modelname', type=str, default=None)
    return parser.parse_args()
args = get_arguments()

metric_save_path = '../metric-py'
# ds_name='mit1003';
ds_name = args.dsname
debug = args.debug
color = Colored()
## salicon
if ds_name == 'salicon':
    sal_base = '/data/sunnycia/SaliencyDataset/Image/SALICON/DATA/train_val/val2014/saliency/';
    dens_dir = '/data/sunnycia/SaliencyDataset/Image/SALICON/DATA/train_val/val2014/density';
    fixa_dir = '/data/sunnycia/SaliencyDataset/Image/SALICON/DATA/train_val/val2014/fixation';
elif ds_name == 'mit1003':
    ## mit1003
    sal_base = '/data/sunnycia/SaliencyDataset/Image/MIT1003/saliency'
    dens_dir = '/data/sunnycia/SaliencyDataset/Image/MIT1003/ALLFIXATIONMAPS'
    fixa_dir = '/data/sunnycia/SaliencyDataset/Image/MIT1003/fixPts'
elif ds_name == 'nctu':
    sal_base = '/data/sunnycia/SaliencyDataset/Image/NCTU/AllTestImg/saliency'
    dens_dir = '/data/sunnycia/SaliencyDataset/Image/NCTU/AllFixMap/sigma_52'
    fixa_dir = '/data/sunnycia/SaliencyDataset/Image/NCTU/AllFixPtsMap/FixPtsMap_allfixs'
elif ds_name == 'msu':
    sal_base = '/data/sunnycia/SaliencyDataset/Video/MSU/saliency'
    dens_dir = '/data/sunnycia/SaliencyDataset/Video/MSU/density/sigma32_allinone'
    fixa_dir = '/data/sunnycia/SaliencyDataset/Video/MSU/fixation/mat_allinone'
elif ds_name == 'videoset':
    sal_base = '/data/sunnycia/SaliencyDataset/Video/VideoSet/saliency_map'
    dens_dir = '/data/sunnycia/SaliencyDataset/Video/VideoSet/All_in_one/density'
    fixa_dir = '/data/sunnycia/SaliencyDataset/Video/VideoSet/All_in_one/fixation'

# elif ds_name == 'nus':
#     sal_base = '/data/sunnycia/SaliencyDataset/Image/NUS/saliency'
#     dens_dir = '/data/sunnycia/SaliencyDataset/Image/NUS/Density'
#     fixa_dir = '/data/sunnycia/SaliencyDataset/Image/NUS/Fixation'

if ds_name not in sal_base.lower():
    print "Caution the dataset version"
    exit()

def videoset_name_sort(imagepath):
    basename = os.path.basename(imagepath)
    video_index = int(basename.split('_')[0].replace('videoSRC', ''))
    frame_index = int(basename.split('.')[0].split('_')[-1])

    return video_index * 1000 + frame_index

evaluation_list = os.listdir(sal_base)
for evaluation in evaluation_list:
    # if evaluation == 'v1':
    #     continue
    if args.modelname is not None:
        evaluation=args.modelname

    output_path = os.path.join(metric_save_path, ds_name+'_'+evaluation+'.mat')
    print color.red(output_path)
    if os.path.isfile(output_path):
        print output_path, "already exists, skip.."
        continue
    print "Info: evaluating", color.red(evaluation)
    # if already done:
    #     pass this file

    # saliency_map_list = glob.glob(os.path.join(sal_base, "*.*"))
    saliency_map_list = glob.glob(os.path.join(sal_base, evaluation, "*.*"))
    density_map_list = glob.glob(os.path.join(dens_dir,  "*.*"))
    fixation_map_list = glob.glob(os.path.join(fixa_dir,  "*.*"))
    if ds_name =='videoset':
        saliency_map_list.sort(key=videoset_name_sort)
        density_map_list.sort(key=videoset_name_sort)
        fixation_map_list.sort(key=videoset_name_sort)
    else:
        saliency_map_list.sort()
        density_map_list.sort()
        fixation_map_list.sort()
    print len(saliency_map_list), len(density_map_list), len(fixation_map_list)
    assert len(saliency_map_list)==len(density_map_list) and len(density_map_list) == len(fixation_map_list)

    ## compute other map

    length = len(saliency_map_list)
    saliency_score_cc = np.zeros((length))
    saliency_score_sim = np.zeros((length))
    saliency_score_jud = np.zeros((length))
    saliency_score_bor = np.zeros((length))
    saliency_score_sauc = np.zeros((length))
    saliency_score_emd = np.zeros((length))
    saliency_score_kld = np.zeros((length))
    saliency_score_nss = np.zeros((length))
    if debug:
        length = 2

    for i in range(length):
        saliency_map_path = saliency_map_list[i]
        density_map_path = density_map_list[i]
        fixation_map_path = fixation_map_list[i]

        print os.path.basename(saliency_map_path).split('.')[0], os.path.basename(density_map_path).split('.')[0], os.path.basename(fixation_map_path).split('.')[0]
        assert os.path.basename(saliency_map_path).split('.')[0] == os.path.basename(density_map_path).split('.')[0] == os.path.basename(fixation_map_path).split('.')[0]

        saliency_map = cv2.imread(saliency_map_path, 0).astype(np.float32)
        density_map = cv2.imread(density_map_path, 0).astype(np.float32)
        ##resize saliency map to the same size as density map
        h,w = density_map.shape
        saliency_map = cv2.resize(saliency_map, dsize=(w,h))
        # print saliency_map.shape, density_map.shape;exit()

        if os.path.basename(fixation_map_path).split('.')[-1] == 'mat':
            # print "loading mat file"
            fixation_map = scio.loadmat(fixation_map_path)['fixation']
        else:
            fixation_map = cv2.imread(fixation_map_path, 0)
        # print fixation_map_path
        # print type(fixation_map)

        print "Computing metric for", os.path.basename(saliency_map_path), "..."
        tic()
        saliency_score_cc[i] = CC(saliency_map, density_map)
        # print 'cc', toc(), 
        saliency_score_sim[i] = SIM(saliency_map, density_map)
        # print 'sim', toc(), 
        saliency_score_jud[i] = AUC_Judd(saliency_map, fixation_map, False)
        # print 'jud', toc(), 
        saliency_score_bor[i] = AUC_Borji(saliency_map, fixation_map)
        # saliency_score_sauc[i] = AUC_shuffled(saliency_map, density_map, other_map)
        # saliency_score_emd[i] = EMD(saliency_map, density_map)
        # print 'bor', toc(), 
        saliency_score_kld[i] = KLdiv(saliency_map, density_map)
        # print 'kld', toc(), 
        saliency_score_nss[i] = NSS(saliency_map, fixation_map)
        # print 'nss', toc(), 
        # saliency_score_infoGain[i] = CC(saliency_map, density_map)

        print "INFO: Done!" , toc(), "second passed"
        print color.blue([saliency_score_cc[i], saliency_score_sim[i], saliency_score_jud[i], saliency_score_bor[i], saliency_score_sauc[i], saliency_score_emd[i], saliency_score_kld[i], saliency_score_nss[i]])

    saliency_score = [saliency_score_cc, saliency_score_sim, saliency_score_jud, saliency_score_bor, saliency_score_sauc, saliency_score_emd, saliency_score_kld, saliency_score_nss]

    if not debug == True:
        scio.savemat(output_path, {'saliency_score':saliency_score})
        print output_path, 'saved'
    if args.modelname is not None:
        break